libs/pika/execution_base/include/pika/execution_base/this_thread.hpp

```diff
@@ -199,5 +199,61 @@ namespace pika::util {
                 }
             }
         }
+
+        // yield_while_timeout is similar to yield_while, with the
+        // addition of a timeout parameter. If the timeout is exceeded, waiting
+        // is stopped and the function returns false. If the predicate is
+        // successfully waited for the function returns true.
+        template <typename Predicate>
+        [[nodiscard]] bool
+        yield_while_timeout(Predicate&& predicate, std::chrono::duration<double> timeout,
+            const char* thread_name = nullptr, bool allow_timed_suspension = true)
+        {
+            // Seconds represented using a double
+            using duration_type = std::chrono::duration<double>;
+
+            bool use_timeout = timeout >= duration_type(0.0);
+
+            pika::chrono::detail::high_resolution_timer t;
+
+            if (allow_timed_suspension)
+            {
+                for (std::size_t k = 0;; ++k)
```

Comment:
It's a good question. I think the original idea here was to avoid doing conditionals inside the loop (i.e. deciding whether or not to do `k % 16` in the `yield_k`) but I'm not sure anyone measured if that makes a difference here. I'd suspect it won't make a big difference since we're just waiting here (i.e. we don't necessarily need to get each iteration done as quickly as possible).

However, I'm going to leave it like this in this PR since the other variants have the same problem. We can try to refactor this separately.